crates/accelerate/src/synthesis/clifford/random_clifford.rs

```diff
@@ -0,0 +1,163 @@
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::synthesis::linear::utils::{
+    binary_matmul_inner, calc_inverse_matrix_inner, replace_row_inner, swap_rows_inner,
+};
+use ndarray::{concatenate, s, Array1, Array2, ArrayView2, ArrayViewMut2, Axis};
+use rand::{Rng, SeedableRng};
+use rand_pcg::Pcg64Mcg;
+
+/// Sample from the quantum Mallows distribution.
+fn sample_qmallows(n: usize, rng: &mut Pcg64Mcg) -> (Array1<bool>, Array1<usize>) {
+    // Hadamard layer
+    let mut had = Array1::from_shape_fn(n, |_| false);
+    // Permutation layer
+    let mut perm = Array1::from_shape_fn(n, |_| 0);
+    let mut inds: Vec<usize> = (0..n).collect();
+
+    for i in 0..n {
+        let m = n - i;
+        let eps: f64 = 4f64.powi(-(m as i32));
+        let r: f64 = rng.gen();
+        let index: usize = -((r + (1f64 - r) * eps).log2().ceil() as isize) as usize;
+        had[i] = index < m;
+        let k = if index < m { index } else { 2 * m - index - 1 };
+        perm[i] = inds[k];
+        inds.remove(k);
+    }
+    (had, perm)
+}
+
+/// Add symmetric random boolean value to off diagonal entries.
+fn fill_tril(mut mat: ArrayViewMut2<bool>, rng: &mut Pcg64Mcg, symmetric: bool) {
+    // todo: implement & benchmark optimizations for low-rank matrices
+    let n = mat.shape()[0];
+    for i in 0..n {
+        for j in 0..i {
+            mat[[i, j]] = rng.gen();
+            if symmetric {
+                mat[[j, i]] = mat[[i, j]];
+            }
+        }
+    }
+}
+
+/// Invert a lower-triangular matrix with unit diagonal.
+fn inverse_tril(mat: ArrayView2<bool>) -> Array2<bool> {
+    // todo: implement & benchmark optimizations for low rank.
+    calc_inverse_matrix_inner(mat, false).unwrap()
+}
+
+/// Generate a random Clifford tableau.
+///
+/// The Clifford is sampled using the method of the paper "Hadamard-free circuits
+/// expose the structure of the Clifford group" by S. Bravyi and D. Maslov (2020),
+/// `https://arxiv.org/abs/2003.09412`__.
+///
+/// The function returns a random clifford tableau.
+pub fn random_clifford_tableau_inner(num_qubits: usize, seed: Option<u64>) -> Array2<bool> {
+    let mut rng = match seed {
+        Some(seed) => Pcg64Mcg::seed_from_u64(seed),
+        None => Pcg64Mcg::from_entropy(),
+    };
+
+    let (had, perm) = sample_qmallows(num_qubits, &mut rng);
+
+    let mut gamma1: Array2<bool> = Array2::from_elem((num_qubits, num_qubits), false);
+    for i in 0..num_qubits {
+        gamma1[[i, i]] = rng.gen();
+    }
+    fill_tril(gamma1.view_mut(), &mut rng, true);
+
+    let mut gamma2: Array2<bool> = Array2::from_elem((num_qubits, num_qubits), false);
+    for i in 0..num_qubits {
+        gamma2[[i, i]] = rng.gen();
+    }
+    fill_tril(gamma2.view_mut(), &mut rng, true);
```

Comment:
Another idea for improvement would be to have `fill_tril` do the random diagonal, and avoid constructing the matrix just to rewrite almost all entries again 🙂 E.g.
```
fn get_tril(rng: ... , symmetric: bool, random_diag: bool) -> Array2<bool>
```
where `random_diag=false` sets the diagonal to `true` (for the `delta` matrices) and `random_diag=true` creates the random diagonal for the `gamma` matrices

Reply:
Technically speaking only half of the entries (when `symmetric` is False). :) I will use the same excuse that I tried to keep the code as close as possible to the original. In any case I don't think this makes much runtime difference (compared to the matrix multiplication between `table1` and `table`).